### DISCUSSION TITLE
Nested use of @ConfigurationBuilder; Another config object inside ConfigurationBuilder

### DISCUSSION THREAD

**Discussion by subash89** [micronaut-projects/micronaut-core#8192] [General] [answered]
Hi,
This ConfigurationBuilder example works fine. (AWSConfig class is outside AppConfig, so using  ConfigurationBuilder as follows).
My question is, I need a new Object field of Metadata in AWSConfig class. (commented out). 

How AWSConfig.Builder  can have another object Metadata ? (cannot think of how to use a nested ConfigurationBuilder or ConfigurationProperties for this)

```
@ConfigurationProperties("app.main")
public class AppConfig {

    @ConfigurationBuilder(configurationPrefix = "aws")
   private AWSConfig.Builder builder = AWSConfig.builder();

   public void setBuilder( AWSConfig.Builder builder ){
    this.builder=builder;
  }

} 
```

```
public class AWSConfig {
    private String region;

  // private Metadata metadata;
     
   public AWSConfig(String region){
     this.region = region;
   }  

  public static class Builder{

      private String region;
      
       public Builder setRegion(String region){
             this.region=region;
             return this;
       }

      public AWSConfig build(){
         return new AWSConfig(region);
      }  
 }

      public static Builder builder(){
        return new Builder();
      }

}
```

```

public class Metadata {
   private String account;
   
    public String getAccount(){
    return account;
     }

     public void setAccount(String account){
     this.account=account;
     }

}
```

**Comment by subash89** [comment]
Changed the subject also a bit if I am not making it clear. All examples I see with ConfigurationBuilder is to build something with java literals. But I have another object in that Builder as in above example.(commented out Metadata object). Is this even possible ?

**Comment by dstepanov** [comment]
There is an example in the configuration of how to use a custom converter https://docs.micronaut.io/latest/guide/#customTypeConverter

**Reply by subash89** [reply]
@dstepanov  In this example, the source type(Map) is given as a programatically added linked hash map  at Application run. 
Can type converter pickup something in application.yml

**Reply by dstepanov** [reply]
It should pickup the map of the configuration at that property location

**Reply by subash89** [reply]
it does not(for me). It works only when I manually pass a manually created HashMap to the ApplicationContext.run(Map).
If you tried it, do you have an example for me to look?

**Reply by subash89** [reply]
@dstepanov **Expect following map style yaml block to picked up as a Source type map by the converter.**
```
myconfig: 
    metadata:
      account: abc123
```

**Instead what works now is manually input map**
  ```
      ApplicationContext ctx  = Micronaut.build(args)
                .mainClass(Application.class)
                .properties(new LinkedHashMap<String, Object>() {{
                    put("myconfig.metadata",
                            new LinkedHashMap<String, String>() {{
                                put("account", "abc123");
                            }}
                    );
                }})
                .start();
```

**Reply by subash89** [reply]
@dstepanov I tried source type List<String> and TypeConverter getting invoked and it works. When I use Map like above, my typeconverter not getting invoked. (verified by, debug pointer not getting hit).

**Comment by subash89** [comment, marked as answer]
From my observation, this works directly inside an object with @ConfigurationProperties() and mapped into a map of objects. Not nested inside child objects.
